File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bruteforce/Problem14500.java
package com.keunwon.algorithm.bruteforce;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Problem14500 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final int[][] moves = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

    private static int N, M;
    private static int[][] map;

    private static boolean[][] visited;
    private static int max = Integer.MIN_VALUE;

    public static void main(String[] args) throws IOException {
        input();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                dfs(i, j, 0, 0);
                otherShape(i, j);
            }
        }

        System.out.println(max);
    }

    private static void dfs(int x, int y, int sum, int depth) {
        if (depth == 4) {
            max = Math.max(max, sum);
            return;
        }

        for (int[] move : moves) {
            final int nx = x + move[0];
            final int ny = y + move[1];

            if (validMap(nx, ny) && !visited[nx][ny]) {
                visited[nx][ny] = true;
                dfs(nx, ny, sum + map[nx][ny], depth + 1);
                visited[nx][ny] = false;
            }
        }
    }

    private static void otherShape(int x, int y) {
        int min = Integer.MAX_VALUE;
        int sum = map[x][y];
        int count = 0;

        for (int[] move : moves) {
            final int nx = x + move[0];
            final int ny = y + move[1];

            if (!validMap(nx, ny)) { continue; }

            final int num = map[nx][ny];
            sum += num;
            min = Math.min(min, num);
            count++;
        }

        if (count < 3) { return; }
        else if (count == 4) { sum -= min; }
        max = Math.max(max, sum);
    }

    private static boolean validMap(int x, int y) {
        return 0 <= x && x < N && 0 <= y && y < M;
    }

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        map = new int[N][M];
        visited = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }
}

File: algorithm/kotlin-algorithm/baekjoon/src/main/kotlin/algorithm/baekjoon/Problem14500.kt
package algorithm.baekjoon

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

/**
 * <p>
 * 이름: 테트로미노
 * 난이도: 골드-4
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/14500">테트로미노 (골드-4)</a>
 **/
class Problem14500 {
    private lateinit var board: Array<IntArray>
    private lateinit var visited: Array<BooleanArray>

    private var result = 0

    fun solution(board: Array<IntArray>): Int {
        this.board = board
        this.visited = Array(board.size) { BooleanArray(board[0].size) }

        for (i in board.indices) {
            for (j in board[i].indices) {
                dfs(0, i, j, 0)
            }
        }
        return result
    }

    private fun dfs(depth: Int, r: Int, c: Int, sum: Int) {
        if (depth == 4) {
            result = result.coerceAtLeast(sum)
            return
        }

        for ((mr, mc) in moves) {
            val nr = r + mr
            val nc = c + mc

            if (nr in board.indices && nc in board[0].indices && !visited[nr][nc]) {
                if (depth == 2) {
                    visited[nr][nc] = true
                    dfs(depth + 1, r, c, sum + board[nr][nc])
                    visited[nr][nc] = false
                }

                visited[nr][nc] = true
                dfs(depth + 1, nr, nc, sum + board[nr][nc])
                visited[nr][nc] = false
            }
        }
    }

    companion object {
        private val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)
    }
}

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val board = Array(n) {
        val st = StringTokenizer(readLine())
        IntArray(m) { st.nextToken().toInt() }
    }

    println(Problem14500().solution(board))
}
